### ISSUE TITLE
Please disable SignatureService service and process on NOT Huawei devices

### ISSUE THREAD

**Issue by ale5000-git** [microg/GmsCore#2040]
@mar-v-in
Please disable SignatureService service and process on NOT Huawei devices (on both GmsCore and FakeStore), they use memory even if not much and old devices have few memory.

**Comment by mar-v-in** [comment]
They shouldn't be using memory, but I agree that it's generally a good idea to disable the service on devices that don't use it.

**Comment by ale5000-git** [comment]
Isn't possible to keep the code inside the apk but activate it only on Huawei devices?

**Comment by mar-v-in** [comment]
We can't really detect meaningfully if it's a banned (huawei) device or not. Huawei's differentiated signature API does not provide any way to detect if a device supports it or not.  I don't think maintaining a list of supported devices (and OS versions) would be reasonable. We will just have a Huawei and non-Huawei builds and users decide which one to take (also: I plan to add some more customizations specifically for Huawei devices, so having two build flavors is a practical thing to do anyways).

**Comment by ale5000-git** [comment]
An alternative way is also to use `res/values/bools.xml` to make choices.
These can be overriden with overlays without the need to change the apk.

**Comment by mar-v-in** [comment]
You mean using RROs?

**Comment by ale5000-git** [comment]
Yes, but a ROM author can bundle official microG with RROs in the system partition and also flashable zips can do it.

Maybe there can be a RROs cofigurable settings as "default" and then an option in microG settings to allow enable/disable it as use choice.

**Comment by mar-v-in** [comment]
I'm generally looking for solutions that can be applied without modifying the ROM. Also I don't think any of the devices and OS versions that support Huawei's differentiated signature API are custom ROMs or can be modified using a flashable zip.

In general, for custom ROM builders and flashable zips we do have microg.xml to override default configuration. This could also be done using RROs and I'm not against doing that (although I don't see the immediate benefit).

**Comment by ale5000-git** [comment]
On recent version of Android there is also another type of overlay that can be enabled/disabled via adb; but currently I don't remember the exact details.

**Comment by ale5000-git** [comment]
microg.xml doesn't works in many of my use cases.
Since I install microG also on a running system it doesn't apply it.

**Comment by ale5000-git** [comment]
The microg.xml doesn't works in many of my use cases.
Since I install microG also on a running system it doesn't apply it.

An idea to improve:
```
microg.xml should be read only the first time of if the app data of microG are deleted.
It should store in its data if the xml was already read or not and if not try to read it in all these cases:
  <action android:name="android.intent.action.BOOT_COMPLETED" />
  <action android:name="android.intent.action.MY_PACKAGE_REPLACED" />
  <action android:name="android.intent.action.PACKAGE_RESTARTED" />
```

**Comment by mar-v-in** [comment]
The `microg.xml` isn't read and imported at any specific point. It overwrites the default value of the configuration options. If you add a `microg.xml` to the system and microG was already installed before, the user might have done configurations already and in that case, the configurations done by the user are not overwritten. If microG was not installed before `microg.xml` is added to the system or the user resets the app data, there is no user configuration and thus all settings in `microg.xml` become active automatically. As long as the user does not change an option in microG settings, `microg.xml` will continue to be the configuration source for that option, thus any future changes to `microg.xml` will be applied as well (including removing options in `microg.xml` which means the default option is changed back to microG default).

There also is the provisioning service which allows system apps to reconfigure microG at runtime. This can be used from setup tools or even from the custom ROMs settings app.

**Comment by ale5000-git** [comment]
On the KitKat emulator with root I place microg.xml and apps in the system partition on a running system and thet work without restart but microg.xml is never used.

**Comment by mar-v-in** [comment]
`microg.xml` must be present at the time microG starts (we don't assume that system partition is modified at runtime). I never tested this on KitKat though, so it might just be broken on KitKat.

Device registration, if enabled, is attempted at boot, package reinstall or on network change, if it hasn't happened in the last 12 hours. If `microg.xml` enabled device registration, it is already enabled at boot and thus device registration happens. The receiver to trigger device registration is configured in the manifest and thus notified even if device registration is disabled (however, it will neither do a device registration nor register a network change listener if device registration is disabled, so it effectively does nothing).

**Comment by ale5000-git** [comment]
@mar-v-in: Can't it also attempt device registration on `android.intent.action.MY_PACKAGE_REPLACED` and on microG settings cleared?

**Comment by mar-v-in** [comment]
That's what I meant with package reinstall. Not sure if it's also triggered on data removal (or, if you can even listen on that)

**Comment by ale5000-git** [comment]
Maybe adding also these?
[ACTION_PACKAGE_DATA_CLEARED](https://developer.android.com/reference/android/content/Intent#ACTION_PACKAGE_DATA_CLEARED)
[ACTION_PACKAGES_UNSUSPENDED](https://developer.android.com/reference/android/content/Intent#ACTION_PACKAGES_UNSUSPENDED)

**Comment by ale5000-git** [comment]
Also if I `am force-stop` microG, even using this `am broadcast -a 'org.microg.gms.gcm.FORCE_TRY_RECONNECT' -n 'com.google.android.gms/org.microg.gms.gcm.TriggerReceiver'` can't bring it back GCM to life.

**Comment by mar-v-in** [comment]
[ACTION_PACKAGE_DATA_CLEARED](https://developer.android.com/reference/android/content/Intent#ACTION_PACKAGE_DATA_CLEARED) says:
> Note that the cleared package does not receive this broadcast.

After a force stop, all receivers in manifest are disabled until next start of app or system reboot.

Also I checked, the method to read preferences from a system file was only added in Android 7, that's why `microg.xml` does not work on KitKat.

**Comment by ale5000-git** [comment]
Is it possible to listen to an intent like: `org.microg.gms.ENSURE_HEALTH`
that when called microG check that all processed/services that need to be running are executed (in case something crashed or got closed by the battery saving apps), that the default values are properly set and that device registration is attempted if needed?

**Comment by ale5000-git** [comment]
@mar-v-in: The device profile is read correctly on KitKat, why not using a similar code even for `microg.xml`?
